File: src/main/java/io/github/codex/mccontrol/ChatHistory.java
package io.github.codex.mccontrol;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

final class ChatHistory {
    private static final int MAX_KEYS = 4096;
    private static final int MAX_MESSAGES = 1024;

    private final Set<String> seenKeys = new HashSet<>();
    private final Deque<String> keyOrder = new ArrayDeque<>();
    private final Deque<Map<String, Object>> messages = new ArrayDeque<>();

    private long nextSequence = 1L;

    public synchronized void record(List<ChatMessageSnapshot> currentMessages) {
        for (ChatMessageSnapshot currentMessage : currentMessages) {
            if (!seenKeys.add(currentMessage.uniqueKey())) {
                continue;
            }

            keyOrder.addLast(currentMessage.uniqueKey());
            while (keyOrder.size() > MAX_KEYS) {
                String oldestKey = keyOrder.removeFirst();
                seenKeys.remove(oldestKey);
            }

            Map<String, Object> payload = new LinkedHashMap<>();
            payload.put("sequence", nextSequence++);
            payload.put("addedTime", currentMessage.addedTime());
            payload.put("text", currentMessage.text());
            payload.put("tag", currentMessage.tag());
            messages.addLast(payload);
            while (messages.size() > MAX_MESSAGES) {
                messages.removeFirst();
            }
        }
    }

    public synchronized Map<String, Object> snapshot(long sinceExclusive, int limit, List<String> recentTyped) {
        List<Map<String, Object>> selected = new ArrayList<>();
        int safeLimit = Math.max(1, limit);

        for (Map<String, Object> message : messages) {
            long sequence = ((Number) message.get("sequence")).longValue();
            if (sequence > sinceExclusive) {
                selected.add(message);
            }
        }

        int fromIndex = Math.max(0, selected.size() - safeLimit);
        List<Map<String, Object>> trimmed = new ArrayList<>(selected.subList(fromIndex, selected.size()));

        Map<String, Object> payload = new LinkedHashMap<>();
        payload.put("ok", Boolean.TRUE);
        payload.put("messages", trimmed);
        payload.put("recentTyped", recentTyped);
        payload.put("latestSequence", nextSequence - 1);
        return payload;
    }

    record ChatMessageSnapshot(int addedTime, String text, String tag) {
        public String uniqueKey() {
            return addedTime + "|" + tag + "|" + text;
        }
    }
}

File: lan-bot/src/chat-history.js
"use strict";

class ChatHistory {
  constructor(maxMessages = 500) {
    this.maxMessages = maxMessages;
    this.startedAt = Date.now();
    this.nextSequence = 1;
    this.messages = [];
    this.recentTyped = [];
  }

  add(text, tag = "System") {
    if (typeof text !== "string" || text.length === 0) {
      return null;
    }

    const entry = {
      sequence: this.nextSequence++,
      addedTime: Date.now() - this.startedAt,
      text,
      tag
    };

    this.messages.push(entry);
    if (this.messages.length > this.maxMessages) {
      this.messages.splice(0, this.messages.length - this.maxMessages);
    }
    return entry;
  }

  addTyped(text) {
    if (typeof text !== "string" || text.length === 0) {
      return;
    }
    this.recentTyped.push(text);
    if (this.recentTyped.length > 20) {
      this.recentTyped.splice(0, this.recentTyped.length - 20);
    }
  }

  get(since = -1, limit = 50) {
    const safeLimit = Number.isFinite(limit) ? Math.max(1, Math.min(200, Math.floor(limit))) : 50;
    const selected = this.messages.filter((entry) => entry.sequence > since);
    const trimmed = selected.slice(Math.max(0, selected.length - safeLimit));
    const latest = this.messages.length > 0 ? this.messages[this.messages.length - 1].sequence : 0;

    return {
      ok: true,
      messages: trimmed,
      recentTyped: this.recentTyped.slice(),
      latestSequence: latest
    };
  }
}

module.exports = ChatHistory;
